### PR TITLE
Autodetect Kafka serializer/deserializer with Reactive Messaging

### LINKED ISSUE #16544: Improve serializer / deserializer detection the Kafka connector
## Description

At the moment, the user must specify the key and value serializer/deserializer when using reactive messaging with Kafka:

```
mp.messaging.foo.incoming.value.deserializer=....

mp.messaging.bar.outgoing.value.serializer=....
```

(for key, if not set "String" is used).

In Quarkus we can detect the right value by inspecting the code:

* methods having an `@Incoming("foo")
* methods having an `@Outgoing("bar")
* injection of the `@Channel("foo") (Multi/Publisher<X>, Multi/Publisher<Message<X>>)
* injection of the `@Channel("bar") Emitter<X> or Emitter<Message<X>> or MutinyEmitter

There is a few cases to consider:

* method can have many potential signatures - See https://docs.google.com/spreadsheets/d/11cNwX8eXKTKjoF_amXTPLU1-DfrYxwu4VLnEideBy7Q/edit?usp=sharing for details
* `Record<K, V>` can be used, and in this case, it's the K/V that are interesting
* It needs to handle X but also `Message<X>`

Of course, if the configuration contains an explicit deserializer or serializer, this one should be used.

## The Avro case

Avro is a bit different.
If the detected type is X and X is an Avro-generated class, Avro _serde_ should be used. Depending on if Confluent or Apicurio is in the classpath the right class names should be used.

### PR DESCRIPTION
**Pull request by Ladicek**
Resolves #16544

### CODE REVIEW

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java [hunk 1]
```diff
@@ -26,4 +40,492 @@ public void build(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
...
+            case ARRAY:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    // ---
+
+    // TODO can we use Map.ofEntries(...) now that we're on 11?
```
Comment (Ladicek) [inline extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java:415]: I'd really like an answer here :-)

Comment (geoand) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java:415]: Yes

Comment (Ladicek) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java:415]: Gonna change that then, thanks!

Comment (gsmet) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java]: As long as you don't plan on getting a patch backported, you can use Java 11 goodness.

Comment (Ladicek) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/SmallRyeReactiveMessagingKafkaProcessor.java]: I don't think we do. Do we, @cescoffier?

**Comment by Ladicek** [comment]
I just found that in addition to `V`, `Record<K, V>` and `KafkaRecord<K, V>`, we also support `ConsumerRecord<K, V>`. Gonna add support for that as well.

**Comment by Ladicek** [comment]
Done.

File: docs/src/main/asciidoc/kafka.adoc [hunk 2]
```diff
@@ -629,6 +633,76 @@ public class FruitResource {
...
+* `double` and `java.lang.Double`
+* `byte[]`
+* `java.lang.String`
+* `java.util.UUID`
+* `java.nio.ByteBuffer`
+* `org.apache.kafka.common.utils.Bytes`
+* `io.vertx.core.buffer.Buffer`
+* `io.vertx.core.json.JsonObject`
+* `io.vertx.core.json.JsonArray`
+* classes generated from Avro schemata, if Confluent or Apicurio _serde_ is present
```
Comment (cescoffier) [inline docs/src/main/asciidoc/kafka.adoc:696]: Schema?

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java [hunk 3]
```diff
@@ -377,7 +377,6 @@ private void handleAvro(BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
                     "io.apicurio.registry.serde.DefaultSchemaResolver",
                     "io.apicurio.registry.serde.DefaultIdHandler",
                     "io.apicurio.registry.serde.Legacy4ByteIdHandler",
-                    "io.apicurio.registry.serde.DefaultSchemaResolver",
```
Comment (cescoffier) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:380]: What's the reason? Is this class gone?

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            }
+        }
+
+        return hasApicurio2;
+    }
+
+    boolean hasJackson() {
+        if (hasJackson == null) {
+            try {
+                Class.forName("com.fasterxml.jackson.databind.ObjectMapper", false,
```
Comment (cescoffier) [inline extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java:94]: Something I fixed yesterday. Kafka now unconditionally depends on Jackson.

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+import org.jboss.jandex.Type;
+
+import io.smallrye.reactive.messaging.kafka.KafkaConnector;
+
+class DefaultSerdeDiscoveryState {
+    private final IndexView index;
+
+    private final Map<String, Boolean> isKafkaConnector = new HashMap<>();
+
+    private Boolean hasConfluent;
```
Comment (cescoffier) [inline extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java:23]: any reason to use objects?

**Review by cescoffier — APPROVED** [review]
LGTM, I've added a few minor comments.

File: docs/src/main/asciidoc/kafka.adoc (hunk 2 above)
Comment (Ladicek) [reply docs/src/main/asciidoc/kafka.adoc:696]: I wanted to use plural, and both "schemas" and "schemata" are apparently OK, but I can go back to singular.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java (hunk 3 above)
Comment (Ladicek) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:380]: It's a duplicate line, see 4 lines above.

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java (hunk 5 above)
Comment (Ladicek) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java:23]: It's a cache. Goes from unknown (`null`) to either `true` or `false`.

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java (hunk 4 above)
Comment (Ladicek) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java:94]: Hmm, does it have to, or is it a convenience?

File: docs/src/main/asciidoc/kafka.adoc (hunk 2 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/kafka.adoc:696]: Oh I didn't know the word! I would go with 'schemas'. I don't believe lots of people will understand 'schematas'.

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java (hunk 4 above)
Comment (cescoffier) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java:94]: It has to. Kafka is now using Jackson to handle authentication, and other stuff too.

File: docs/src/main/asciidoc/kafka.adoc (hunk 2 above)
Comment (geoand) [reply docs/src/main/asciidoc/kafka.adoc:696]: `schematas` is far too greek to be generally understandable 😎

Comment (Ladicek) [reply docs/src/main/asciidoc/kafka.adoc:696]: It's like indexes vs. indices :-) Alright, I'll change it to "schemas".

File: extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java (hunk 4 above)
Comment (Ladicek) [reply extensions/smallrye-reactive-messaging-kafka/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/kafka/deployment/DefaultSerdeDiscoveryState.java:94]: Interesting. Alright, I'll simplify this.

**Review by cescoffier — APPROVED** [review]

**Comment by cescoffier** [comment]
Thanks!

**Pull request merged**